src/auth/mod.rs

```diff
@@ -0,0 +1,171 @@
+pub struct HawkPayload {
+    expires: f64,
+    node: String,
+    salt: String,
+    uid: u64,
+}
+
+impl HawkPayload {
+    fn new(
+        header: &str,
+        method: &str,
+        path: &str,
+        host: &str,
+        port: u16,
+        settings: &Settings,
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        if &header[0..5] != "Hawk " {
+            return Err(AuthError);
+        }
+
+        let header: HawkHeader = header[5..].parse()?;
+        let id = header.id.as_ref().ok_or(AuthError)?;
+
+        let payload = HawkPayload::extract_and_validate(id, &settings.master_secret, expiry)?;
+
+        let token_secret = hkdf_expand_32(
+            format!("services.mozilla.com/tokenlib/v1/derive/{}", id).as_bytes(),
+            Some(payload.salt.as_bytes()),
+            &settings.master_secret,
+        )?;
+        let token_secret = base64::encode_config(&token_secret, base64::URL_SAFE);
+
+        let request = RequestBuilder::new(method, host, port, path).request();
+        if request.validate_header(
+            &header,
+            &Key::new(token_secret.as_bytes(), &ring::digest::SHA256),
+            Duration::weeks(52),
+        ) {
+            Ok(payload)
+        } else {
+            Err(AuthError)
+        }
+    }
+
+    fn extract_and_validate(
+        id: &str,
+        master_secret: &[u8],
+        expiry: u64,
+    ) -> AuthResult<HawkPayload> {
+        let decoded_id = base64::decode_config(id, base64::URL_SAFE)?;
+        let payload_length = decoded_id.len() - 32;
+        let payload = &decoded_id[0..payload_length];
+        let signature = &decoded_id[payload_length..];
+
+        let signing_secret = hkdf_expand_32(
+            b"services.mozilla.com/tokenlib/v1/signing",
+            None,
+            master_secret,
+        )?;
+
+        let mut expected_signature: Hmac<Sha256> = Hmac::new_varkey(&signing_secret)?;
+        expected_signature.input(&payload);
+        expected_signature.verify(signature)?;
+
+        let payload: HawkPayload = serde_json::from_slice(payload)?;
+
+        if (payload.expires.round() as u64) > expiry {
+            Ok(payload)
+        } else {
+            Err(AuthError)
+        }
+    }
+}
+
+impl<S> FromRequest<S> for HawkPayload {
+    type Config = Settings;
+    type Result = AuthResult<HawkPayload>;
+
+    /// Extract a HAWK payload from an actix request object.
+    fn from_request(request: &HttpRequest<S>, settings: &Self::Config) -> Self::Result {
+        HawkPayload::new(
+            request
+                .headers()
+                .get("authorization")
+                .ok_or(AuthError)?
+                .to_str()?,
+            request.method().as_str(),
+            request.path(),
+            request.uri().host().ok_or(AuthError)?,
+            request.uri().port().ok_or(AuthError)?,
+            settings,
+            // Allow a generous amount of client skew for expiry
+            Utc::now().timestamp() as u64 + YEAR,
```

Comment:
This `expiry` date applies to the `expires` timestamp that is decoded from inside the Hawk id JSON payload, right?  That timestamp is generated server-side by the tokenserver (and is covered by the HMAC signature check in `extract_and_validate`) so we do not need to allow for client-side clock skew in that check.  We should assume that the servers have well-synchronized clocks.

As written, this would allow tokenserver tokens to be valid for up to a year longer than intended, which we definitely don't want.

(The timestamp from the hawk header *is* client-generated and we can allow some generous clock skew on that one).

Reply:
Oh, my bad. I only added this after the suggestion in https://github.com/mozilla-services/syncstorage-rs/pull/20#discussion_r215839404.